_posts/2016-10-06-BDD-tests-for-Hadoop-with-Cucumber-part-I.md

```diff
@@ -0,0 +1,310 @@
+for example abstraction to Amazon S3, so we can use S3 buckets as input/output folders for processing jobs, 
+but in the same time we allow to use regular file systems on developer machines in development and test purposes.
+
+As any other piece of software it should be tested before using in production, especially because logical mistakes in processing a lot of data cost a much. 
+Unfortunately there are no much information about BDD testing the Hadoop and Hive jobs. 
+So i decide to write how we do it in our Cascading project here, in Intelliarts.
+
+<!-- more -->
+
+# What do we have
+
+Assume we have a lot of data about phone calls in bunch of files. File contain lines with following format: `Id,country,time,duration` 
+Where `id` is call unique id, `country` is a caller country, `time` is a time when call start (timestamp) and `duration` is a call duration in milliseconds. 
+
+For example: 
+```
+0000,UA,1433998201,60000
+0001,US,1433998201,30000
+0001,GB,1433998301,30000
+...
+```
+
+# What we want to do?
+
+Let's sort calls from these records by country (each country will be in separate folder) and change the format of records - replace delimiters from `,` to tab symbol `\t`. 
+From the date of previous example so we should get the folders `US`, `GB`, `UA` and `UA` will contain file with records looks like this:
+
+```
+id	country	time	duration
+0000	UA	1433998201	60000
+```
+
+And, we want to be sure that we did any mistakes during implementation, so we want to test the results. 
+What we will use:  
+* Hadoop + Cascading for data processing
+* Gradle for build management
+* Cucumber for tests
+* Chunk Templates  https://github.com/tomj74/chunk-templates for test data fixture generation
+* Docker for environment isolation
+
+First of all we should isolate our testing environment and make it portable between dev machines and CI. We use Docker for this propose. 
+Inside the container we will:
+* Install Java
+* Install Hadoop
+* Install Gradle
+* Add codebase
+Here is Dockerfile content:
+
+{% highlight groovy %}
+FROM ubuntu:14.04
+
+MAINTAINER Bogdan Frankovskyi <bogdan@savvyclutch.com>
+
+RUN apt-get update ; apt-get upgrade -y
+
+### INSTALL JAVA
+RUN apt-get install -y default-jdk wget software-properties-common
+
+### ADD HADOOP
+RUN mkdir -p /usr/local/hadoop
+RUN cd /tmp/ ; wget http://mirrors.sonic.net/apache/hadoop/common/hadoop-2.6.0/hadoop-2.6.0.tar.gz ; tar xzf hadoop-2.6.0.tar.gz ; mv hadoop-2.6.0/* /usr/local/hadoop
+
+### ADD GRADLE
+RUN add-apt-repository ppa:cwchien/gradle
+RUN apt-get update ; apt-get install -y gradle
+
+ENV GRADLE_USER_HOME /cache
+
+#### ADD THE DIRECTORY FOR CODEBASE
+ADD . /opt/bdd_test_hadoop/
+WORKDIR /opt/bdd_test_hadoop/
+
+#### DOWNLOAD AND INSTALL DEPENDENCIES
+RUN gradle dependencies && gradle install
+
+#### RUN TESTS
+
+CMD gradle cucumber
+{% endhighlight %}
+
+Okey, now we have environment. Let’s create project gradle config to download all dependencies for Hadoop application:
```

Comment:
Okey, now we have environment. Let’s create project gradle config to download all dependencies for Hadoop application:
Okey, now we have **the** environment. Let’s create **a** project gradle config to download all dependencies for Hadoop application: